### PR TITLE
Us001 view job roles

### PR DESCRIPTION
**Pull request by johnmcg23**
This PR references ticket US001, this user story states: 

_As a Kainos employee
I want to view job roles within Kainos 
So that I can see all the job roles available in Kainos_

In this PR I have written a test to test the getAllJobRoles() method in the jobRoleService class, mocking the JobRoleDAO class:
Screenshot 2023-12-14 at 22 09 04
_
_
_

I have also hit the /job-role-list endpoint to ensure it works, this returned a 200 code indicating everything was 👍:
Screenshot 2023-12-14 at 22 27 16

_
_
_
**Code changes**: 🔴**_This will have to be carried through into the other tickets_** 🔴
In this PR I have changed around parameters for the getAllJobRoles() method, this was due to being able to test the JobService class.

### CODE REVIEW

**Review by autumncraig — APPROVED** [review]

File: FlashFeetAPI/src/main/java/org/example/db/JobRoleDao.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+public class JobRoleDao {
+    private DatabaseConnector databaseConnector = new DatabaseConnector();
+    public List<JobRole> getAllJobRoles(Connection connection) throws SQLException {
+
+        // Connection c = databaseConnector.getConnection();
```
Comment (jkelso662) [inline FlashFeetAPI/src/main/java/org/example/db/JobRoleDao.java:16]: Commented out code shouldn't be commited!

File: FlashFeetAPI/src/main/java/org/example/exception/DatabaseConnectionException.java [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package org.example.exception;
+
+public class DatabaseConnectionException extends Throwable {
```
Comment (jkelso662) [inline FlashFeetAPI/src/main/java/org/example/exception/DatabaseConnectionException.java:3]: Should this not be implimented?

File: FlashFeetAPI/src/main/java/org/example/exception/JobRoleLengthException.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package org.example.exception;
+
+public class JobRoleLengthException extends Throwable {
```
Comment (jkelso662) [inline FlashFeetAPI/src/main/java/org/example/exception/JobRoleLengthException.java]: This also needs implemented

File: FlashFeetAPI/src/test/java/org/example/api/service/JobRoleServiceTests.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+//
+//                List<JobRole> listOfJobRoles = Arrays.asList(
+//                        Mockito.mock(JobRole.class)
+//                );
+//
+//                Mockito.when(databaseConnector.getConnection()).thenReturn(c);
+//                Mockito.when(jobRoleDao.getAllJobRoles(c)).thenReturn(new ArrayList<>(listOfJobRoles)); // Create a new list from the original list
+//                assertEquals(listOfJobRoles, jobRoleService.getAllJobRoles());
+//        }
+//}
```
Comment (jkelso662) [inline FlashFeetAPI/src/test/java/org/example/api/service/JobRoleServiceTests.java]: 😬

File: FlashFeetAPI/src/main/java/org/example/trueApplication.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            protected SwaggerBundleConfiguration getSwaggerBundleConfiguration(trueConfiguration configuration) {
+                return configuration.getSwagger();
+            }
+        });
+    }
+
+    @Override
+    public void run(final trueConfiguration configuration,
+                    final Environment environment) {
+        // TODO: implement application
```
Comment (jkelso662) [inline FlashFeetAPI/src/main/java/org/example/trueApplication.java]: Probably don't need this anymore

File: FlashFeetAPI/src/main/java/org/example/trueApplication.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    @Override
+    public String getName() {
+        return "true";
+    }
+
+    @Override
+    public void initialize(final Bootstrap<trueConfiguration> bootstrap) {
+        // TODO: application initialization
```
Comment (jkelso662) [inline FlashFeetAPI/src/main/java/org/example/trueApplication.java]: Or this

**Review by jkelso662 — CHANGES_REQUESTED** [review]
Not bad! Glad you have some tests in there 💯  however there is some commented code in there that can't be allowed in. Remember to review your own PR before assigning to me and Joanne. Also I assume there is still a bunch of stuff to go into Pre-Production because I feel like I've seen your gitignore file added like 4 times now 😆

**Review by johnmcg23 — COMMENTED** [review]
All reviews fixed, ready to merge

**Comment by johnmcg23** [comment]
The .gitignore file being added 4 times is out of safety of my password lmao

File: FlashFeetAPI/src/main/java/org/example/api/JobRoleService.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.List;
+
+public class JobRoleService {
+
+    private JobRoleDao jobRoleDao = new JobRoleDao();
+
+    public DatabaseConnector databaseConnector = new DatabaseConnector();
```
Comment (TaylorJCrawford) [inline FlashFeetAPI/src/main/java/org/example/api/JobRoleService.java:16]: [Possible Future Improvement] - Don't need to create a new instance here. As this is being passed in via your constructor.

File: FlashFeetAPI/src/main/java/org/example/db/DatabaseConnector.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        String user, password, host, name;
+
+        if (conn != null) {return conn;}
+
+        try (FileInputStream propsStream = new FileInputStream("db.properties")){
+
+            Properties props = new Properties();
+            props.load(propsStream);
+
+            user = props.getProperty("user");
```
Comment (TaylorJCrawford) [inline FlashFeetAPI/src/main/java/org/example/db/DatabaseConnector.java]: [Possible Future Improvement]  - Similar to other PRs Environment Variables would be better here.

File: FlashFeetAPI/src/main/java/org/example/db/JobRoleDao.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+public class JobRoleDao {
+    private DatabaseConnector databaseConnector = new DatabaseConnector();
```
Comment (TaylorJCrawford) [inline FlashFeetAPI/src/main/java/org/example/db/JobRoleDao.java:13]: [Possible Improvement] - Don't need to create a new instance here. As you are passing this via your method arguments. If you are interested the best way is to use dependency injection. https://www.freecodecamp.org/news/a-quick-intro-to-dependency-injection-what-it-is-and-when-to-use-it-7578c84fa88f/

File: FlashFeetAPI/src/test/java/org/example/api/validator/JobsRoleValidatorTest.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import org.example.cli.JobRole;
+import org.example.validator.JobRoleValidator;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class JobsRoleValidatorTest {
+    JobRoleValidator jobRoleValidator = new JobRoleValidator();
+    @Test
+    void isValidJobRole_shouldReturnTrue_whenValidJobRole() {
```
Comment (TaylorJCrawford) [inline FlashFeetAPI/src/test/java/org/example/api/validator/JobsRoleValidatorTest.java]: [Possible Improvement] - I think I said this in another review. Having negative tests would be nice to see. As this is currently testing for the happy path. In addition to this boundary testing is also a very useful technique. If your interested have a look at: https://www.tutorialspoint.com/software_testing_dictionary/boundary_testing.htm

File: FlashFeetAPI/dependency-reduced-pom.xml [hunk 11]
```diff
@@ -0,0 +1,153 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (TaylorJCrawford) [inline FlashFeetAPI/dependency-reduced-pom.xml]: File really shouldn't be under version control. If you can, could you remove this.

File: FlashFeetAPI/src/main/java/org/example/api/JobRoleService.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+    public JobRoleService(){
```
Comment (TaylorJCrawford) [inline FlashFeetAPI/src/main/java/org/example/api/JobRoleService.java:18]: [Possible Future Improvement] - Don't really need this. As this provided for you by the Object Class.

**Review by TaylorJCrawford — APPROVED** [review]

**Pull request merged**